File: unipaith-backend/src/unipaith/schemas/calendar.py
"""Spec 16 · Calendar — Pydantic schemas.

The ``CalendarItem`` response mirrors Spec 16 §6 exactly, plus a few
display-only helpers (``subtitle``, ``link``, ``institution_name``,
``recommender_name``, ``editable``) the frontend uses to render rows and
deep-links without re-deriving them.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

CalendarItemType = Literal[
    "interview_live",
    "interview_recorded_window",
    "campus_visit",
    "info_session",
    "portfolio_review",
    "audition",
    "submission_deadline",
    "document_deadline",
    "recommendation_deadline",
    "interview_submission_deadline",
    "deposit_deadline",
    "reminder",
    "work_block",
]

CalendarItemStatus = Literal["scheduled", "completed", "cancelled", "overdue"]

ReminderChannel = Literal["email", "push", "in_app"]


class ReminderSettings(BaseModel):
    lead_time_minutes: int = Field(default=60, ge=0, le=20160)  # up to 14 days
    channels: list[ReminderChannel] = Field(default_factory=lambda: ["in_app"])


class CalendarItem(BaseModel):
    """Spec 16 §6 — one item on the unified admissions timeline."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    type: CalendarItemType
    title: str
    start_at: datetime
    end_at: datetime | None = None
    location: str | None = None
    meeting_link: str | None = None
    application_id: UUID | None = None
    status: CalendarItemStatus = "scheduled"
    notes: str | None = None
    reminder_settings: ReminderSettings | None = None
    # ── display helpers (not part of the §6 contract) ──
    subtitle: str | None = None
    link: str | None = None
    institution_name: str | None = None
    recommender_name: str | None = None
    confirmation_url: str | None = None
    editable: bool = False
    interview_id: UUID | None = None
    can_decline: bool = False
    can_reschedule: bool = False


class ReminderCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    start_at: datetime
    notes: str | None = Field(default=None, max_length=2000)
    application_id: UUID | None = None
    reminder_settings: ReminderSettings | None = None


class WorkBlockCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    start_at: datetime
    end_at: datetime | None = None
    duration_minutes: int | None = Field(default=None, ge=5, le=1440)
    category: str | None = Field(default=None, max_length=30)
    application_id: UUID | None = None
    notes: str | None = Field(default=None, max_length=2000)


class CalendarItemPatch(BaseModel):
    status: CalendarItemStatus | None = None
    notes: str | None = Field(default=None, max_length=2000)
    title: str | None = Field(default=None, max_length=255)
    start_at: datetime | None = None
    end_at: datetime | None = None
    confirmation_url: str | None = Field(default=None, max_length=1000)

File: frontend/src/api/calendar.ts
import apiClient from './client'
import { toArrayData } from './normalize'

// Spec 16 §6 — unified admissions-timeline item.
export type CalendarItemType =
  | 'interview_live'
  | 'interview_recorded_window'
  | 'campus_visit'
  | 'info_session'
  | 'portfolio_review'
  | 'audition'
  | 'submission_deadline'
  | 'document_deadline'
  | 'recommendation_deadline'
  | 'interview_submission_deadline'
  | 'deposit_deadline'
  | 'reminder'
  | 'work_block'

export type CalendarItemStatus = 'scheduled' | 'completed' | 'cancelled' | 'overdue'

export interface ReminderSettings {
  lead_time_minutes: number
  channels: ('email' | 'push' | 'in_app')[]
}

export interface CalendarItem {
  id: string
  type: CalendarItemType
  title: string
  start_at: string
  end_at: string | null
  location: string | null
  meeting_link: string | null
  application_id: string | null
  status: CalendarItemStatus
  notes: string | null
  reminder_settings: ReminderSettings | null
  // display helpers
  subtitle: string | null
  link: string | null
  institution_name: string | null
  recommender_name: string | null
  confirmation_url: string | null
  editable: boolean
}

export interface ReminderCreate {
  title: string
  start_at: string
  notes?: string | null
  application_id?: string | null
  reminder_settings?: ReminderSettings | null
}

export interface WorkBlockCreate {
  title: string
  start_at: string
  end_at?: string | null
  duration_minutes?: number | null
  category?: string | null
  application_id?: string | null
  notes?: string | null
}

export interface CalendarItemPatch {
  status?: CalendarItemStatus
  notes?: string | null
  title?: string | null
  start_at?: string | null
  end_at?: string | null
  confirmation_url?: string | null
}

export const getCalendar = (params?: { from?: string; to?: string }) =>
  apiClient.get('/me/calendar', { params }).then(r => toArrayData<CalendarItem>(r.data))

export const createReminder = (body: ReminderCreate) =>
  apiClient.post('/me/calendar/reminders', body).then(r => r.data as CalendarItem)

export const createWorkBlock = (body: WorkBlockCreate) =>
  apiClient.post('/me/calendar/work-blocks', body).then(r => r.data as CalendarItem)

export const patchCalendarItem = (id: string, body: CalendarItemPatch) =>
  apiClient.patch(`/me/calendar/${encodeURIComponent(id)}`, body).then(r => r.data as CalendarItem)
